### PR TITLE
Boot into hardware setup menu on Vmware

### PR DESCRIPTION
**Pull request by DaanHoogland**
## Description

allow VMs to be started into BIOS on Vmware. This is implemented in the reboot, start and deploy APIs for user VMs.

Related Primate PR: https://github.com/apache/cloudstack-primate/pull/313

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## Screenshots (if appropriate):

image

image

Screenshot 2020-04-27 at 17 13 06

A primate PR has been created with [number 313](https://github.com/apache/cloudstack-primate/pull/313)

## How Has This Been Tested?

this has been manually verified by means of API/old-UI/primate and checking if the boot menu would then be visible or not depending on the flag passed.

### CODE REVIEW

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by svenvogel** [comment]
@DaanHoogland do you think it could be used to extend it for KVM?

**Comment by svenvogel** [comment]
@DaanHoogland maybe i thought of "bios" as boot order. so in VMware i can set them into the bios. in kvm i need an xml entry, but that would have to be implemented too. an example would be the use of PXE. that's what we lack so far. 

```xml
 <bootmenu enable='yes'/> 
```

**Comment by DaanHoogland** [comment]
@svenvogel I think you want to look @bwsw 's latest merged feature, PR #3823. It might give an example of the needed. not sure though.

**Comment by DaanHoogland** [comment]
> @DaanHoogland maybe i thought of "bios" as boot order. so in VMware i can set them into the bios. in kvm i need an xml entry, but that would have to be implemented too. an example would be the use of PXE. that's what we lack so far.
> 
> ```
>  <bootmenu enable='yes'/> 
> ```
> 
> this is the reason why I ask. where do you think this should be implemented?

on second thought, I don't think I understand what you want exactly.

**Comment by DaanHoogland** [comment]
@blueorangutan test matrix

**Comment by DaanHoogland** [comment]
@blueorangutan test matrix

**Comment by DaanHoogland** [comment]
@cl-k-takahashi I saw your interest in this change. Do you care to give it an independent review?

**Comment by svenvogel** [comment]
@DaanHoogland yes, thanks Daan, thats maybe a good solution to use kvm hooks and enable this feature. you are right. i thought from vmware perspective and thought how i get the pxe or boot menu into kvm. maybe its a idea to think about a non hook design like this one for kvm to get "bios" formerly boot options.

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 1]
```diff
@@ -1315,6 +1324,30 @@ public void orchestrateStart(final String vmUuid, final Map<VirtualMachineProfil
...
+        if (params.get(VirtualMachineProfile.Param.BootMode) != null) {
+            msgBuf.append(String.format("Boot Mode: %s ", params.get(VirtualMachineProfile.Param.BootMode)));
+            log = true;
+        }
+        if (params.get(VirtualMachineProfile.Param.BootIntoSetup) != null) {
+            msgBuf.append(String.format("Boot into Setup: %s ", params.get(VirtualMachineProfile.Param.BootIntoSetup)));
+            log = true;
+        }
+        if (log) {
+            s_logger.info(msgBuf.toString());
```
Comment (nvazquez) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:1347]: Minor improvement request: what about removing the boolean variable and init the string buffer empty, so at this point instead of asking for the boolean flag you could check if the generated string is not empty, in that case log it

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java [hunk 2]
```diff
@@ -1723,10 +1723,13 @@ protected StartAnswer execute(StartCommand cmd) {
         String dataDiskController = vmSpec.getDetails().get(VmDetailConstants.DATA_DISK_CONTROLLER);
         String rootDiskController = vmSpec.getDetails().get(VmDetailConstants.ROOT_DISK_CONTROLLER);
         DiskTO rootDiskTO = null;
-        String bootMode = "bios";
+        String bootMode = null;
         if (vmSpec.getDetails().containsKey(VmDetailConstants.BOOT_MODE)) {
             bootMode = vmSpec.getDetails().get(VmDetailConstants.BOOT_MODE);
         }
+        if(null == bootMode) {
```
Comment (nvazquez) [inline plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java:1730]: Missing space

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java [hunk 3]
```diff
@@ -3950,6 +3970,25 @@ protected Answer execute(RebootCommand cmd) {
...
+                s_logger.debug(String.format("configuring VM '%s' to reboot into hardware setup menu.",virtualMachine.getName()));
+            }
+            bootOptions.setEnterBIOSSetup(virtualMachine.isEnterHardwareSetup());
+            vmConfigSpec.setBootOptions(bootOptions);
+            try {
+                if (!vmMo.configureVm(vmConfigSpec)) {
+                    throw new Exception("Failed to configure VM to boot into hardware setup menu: " + vmMo.getName());
+                }
+            } catch (Exception e) {
+                s_logger.error(String.format("failed to reconfigure VM '%s' to boot into hardware setup menu",virtualMachine.getName()),e);
```
Comment (nvazquez) [inline plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java:3987]: We are swallowing the exception here, should we try to stop the process instead?

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 4]
```diff
@@ -969,8 +973,18 @@ private UserVm rebootVirtualMachine(long userId, long vmId) throws InsufficientC
             } catch (Exception ex){
                 throw new CloudRuntimeException("Router start failed due to" + ex);
             }finally {
-                s_logger.info("Rebooting vm " + vm.getInstanceName());
-                _itMgr.reboot(vm.getUuid(), null);
+                if (s_logger.isInfoEnabled()) {
+                    s_logger.info(String.format("Rebooting vm %s%s.", vm.getInstanceName(), enterSetup? " entering hardware setup menu" : " as is"));
+                }
+                Map<VirtualMachineProfile.Param,Object> params = null;
+                if(enterSetup) {
```
Comment (nvazquez) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:980]: Space

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 5]
```diff
@@ -2866,7 +2892,11 @@ public UserVm rebootVirtualMachine(RebootVMCmd cmd) throws InsufficientCapacityE
             throw new InvalidParameterValueException("Unable to find service offering: " + serviceOfferingId + " corresponding to the vm");
         }
 
-        UserVm userVm = rebootVirtualMachine(CallContext.current().getCallingUserId(), vmId);
+        Boolean enterSetup = cmd.getBootIntoSetup();
+        if (! (enterSetup == null || HypervisorType.VMware.equals(vmInstance.getHypervisorType()))) {
```
Comment (nvazquez) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:2896]: For readability can you change it to: enterSetup != null && hypervisor != Vmware?

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/guru/VMwareGuru.java [hunk 6]
```diff
@@ -172,326 +157,59 @@
...
-            List<StoragePoolVO> pools = _storagePoolDao.listByDataCenterId(dataCenterId);
-            for (StoragePoolVO pool : pools) {
-                VMTemplateStoragePoolVO ref = templateStoragePoolDao.findByPoolTemplate(pool.getId(), vm.getTemplateId());
-                if (ref != null && ref.getDownloadState() == VMTemplateStorageResourceAssoc.Status.DOWNLOADED) {
-                    templateInstallPath = ref.getInstallPath();
-                    break;
-                }
-            }
+    @Override public VirtualMachineTO implement(VirtualMachineProfile vm) {
+        vmwareVmImplementer = ComponentContext.inject(new VmwareVmImplementer(this));
```
Comment (nvazquez) [inline plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/guru/VMwareGuru.java:210]: Why not using the @Inject annotation?

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/guru/VMwareGuru.java [hunk 7]
```diff
@@ -686,8 +400,7 @@ public int compare(NicTO arg0, NicTO arg1) {
         return new Pair<Boolean, Long>(Boolean.FALSE, new Long(hostId));
     }
 
-    @Override
-    public boolean trackVmHostChange() {
+    @Override public boolean trackVmHostChange() {
```
Comment (nvazquez) [inline plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/guru/VMwareGuru.java:403]: Just format stuff: these annotations have been moved to the same line as the method's firm all over the class

File: api/src/main/java/org/apache/cloudstack/api/command/user/vm/DeployVMCmd.java [hunk 8]
```diff
@@ -281,15 +284,14 @@ public Long getDomainId() {
                 }
             }
         }
-        if(getBootType() != null){ // export to get
-            if(getBootType() == ApiConstants.BootType.UEFI) {
-                customparameterMap.put(getBootType().toString(), getBootMode().toString());
-            }
+        if (ApiConstants.BootType.UEFI.equals(getBootType())) {
```
[the remark this answers is not in the document]
Comment (DaanHoogland) [reply api/src/main/java/org/apache/cloudstack/api/command/user/vm/DeployVMCmd.java:287]: null won't be equal to ApiConstants.BootType.UEFI so it is implied.

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java (hunk 3 above)
Comment (DaanHoogland) [reply plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java:3987]: no, the rest of the reboot process can still proceed normally and it is an unlikely event that implies an error in Vsphere.

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java (hunk 5 above)
Comment (DaanHoogland) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:2896]: would `enterSetup != null && !HypervisorType.VMware.equals(vmInstance.getHypervisorType())` be acceptable as well?

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java (hunk 1 above)
Comment (DaanHoogland) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:1347]: actually the initial text would have to be added so the logic would replace teh guard boolean with a string concatenation. I prefer this.

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java (hunk 3 above)
Comment (DaanHoogland) [reply plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java:3987]: ok, revisited this and yes it is a valid concern in view of the hard reset that follows in the calling method. I'll have to refactor this to handle a failure to set the parameter on the VM.

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java (hunk 5 above)
Comment (nvazquez) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java]: Sure

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java (hunk 3 above)
Comment (nvazquez) [reply plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java:3996]: Looks good. I can think of a way to test this:
What about adding a config key to control the delay as described here? (bootDelay param) https://vdc-download.vmware.com/vmwb-repository/dcr-public/b50dcbbf-051d-4204-a3e7-e1b618c1e384/538cf2ec-b34f-4bae-a332-3820ef9e7773/vim.vm.BootOptions.html
Will need testing but as it expects a long value, sending a negative delay may cause VM config to throw an exception

Comment (DaanHoogland) [reply plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java:3996]: :) next feature, but good to keep in mind!

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/guru/VMwareGuru.java (hunk 6 above)
Comment (DaanHoogland) [reply plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/guru/VMwareGuru.java:210]: trying to clean it up enough to deal with the cyclic dependencies.

**Comment by DaanHoogland** [comment]
@blueorangutan test matrix

**Comment by DaanHoogland** [comment]
@blueorangutan test centos7 xs71

**Comment by DaanHoogland** [comment]
@blueorangutan test centos7 xenserver-71

**Comment by DaanHoogland** [comment]
the failures in vmware seem related, however on a manual rerun of `test_ssvm`, `test_08_reboot_cpvm` failed instead of `test_07_reboot_ssvm`, as it did in the automated run, giving me the thought this is yet another timing issue.

I'll be investigating `test_nested_virtualization` as well.

**Comment by DaanHoogland** [comment]
last run of test_ssvm actually went fine ???:
 ```
Test List secondary storage VMs ... === TestName: test_01_list_sec_storage_vm | Status : SUCCESS ===
ok
Test List console proxy VMs ... === TestName: test_02_list_cpvm_vm | Status : SUCCESS ===
ok
Test SSVM Internals ... === TestName: test_03_ssvm_internals | Status : SUCCESS ===
ok
Test CPVM Internals ... === TestName: test_04_cpvm_internals | Status : SUCCESS ===
ok
Test stop SSVM ... === TestName: test_05_stop_ssvm | Status : SUCCESS ===
ok
Test stop CPVM ... === TestName: test_06_stop_cpvm | Status : SUCCESS ===
ok
Test reboot SSVM ... === TestName: test_07_reboot_ssvm | Status : SUCCESS ===
ok
Test reboot CPVM ... === TestName: test_08_reboot_cpvm | Status : SUCCESS ===
ok
Test destroy SSVM ... === TestName: test_09_destroy_ssvm | Status : SUCCESS ===
ok
Test destroy CPVM ... === TestName: test_10_destroy_cpvm | Status : SUCCESS ===
ok
Test NFS Version on Secondary Storage mounted properly on SSVM ... SKIP: No NFS version provided in test data

----------------------------------------------------------------------
Ran 11 tests in 705.971s

OK (SKIP=1)
```

**Comment by DaanHoogland** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by borisstoyanov** [comment]
test centos7 vmware-65

**Comment by borisstoyanov** [comment]
@blueorangutan test centos7 vmware-65

**Comment by borisstoyanov** [comment]
@blueorangutan test centos7 vmware-65u2

File: api/src/main/java/com/cloud/agent/api/to/VirtualMachineTO.java [hunk 9]
```diff
@@ -65,6 +65,7 @@
     String uuid;
     String bootType;
     String bootMode;
+    boolean enterHardwareSetup;
```
Comment (rhtyd) [inline api/src/main/java/com/cloud/agent/api/to/VirtualMachineTO.java:68]: @DaanHoogland Is hardware setup same as boot setup?

Comment (DaanHoogland) [reply api/src/main/java/com/cloud/agent/api/to/VirtualMachineTO.java:68]: in hindsite this name should have been something like "bootIntoFirmware". many names have passed the catwalk. on resource level it is called enterHardwareSetup on API it is bootIntoSetup.

**Comment by rhtyd** [comment]
@blueorangutan test centos7 vmware-67u3

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 10]
```diff
@@ -2866,7 +2892,11 @@ public UserVm rebootVirtualMachine(RebootVMCmd cmd) throws InsufficientCapacityE
             throw new InvalidParameterValueException("Unable to find service offering: " + serviceOfferingId + " corresponding to the vm");
         }
 
-        UserVm userVm = rebootVirtualMachine(CallContext.current().getCallingUserId(), vmId);
+        Boolean enterSetup = cmd.getBootIntoSetup();
+        if (enterSetup != null && !HypervisorType.VMware.equals(vmInstance.getHypervisorType())) {
+            throw new InvalidParameterValueException("booting into a hardware setup menu does not make sense on " + vmInstance.getHypervisorType());
```
Comment (andrijapanicsb) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:2897]: I don't like this message, missing capital "B" and in general I don't like it... can we reword it to  "...not supported for ...." or similar?

Comment (DaanHoogland) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:2897]: ah, yes we can but a new re-word PR is needed.

**Comment by rhtyd** [comment]
This or any of the VMware related PRs may have introduced some regression due to consistently failing test `test_07_reboot_ssvm` :( let's fix that soon irrespective of what/who caused this cc @DaanHoogland
